crates/vm/levm/src/precompiles.rs

```diff
@@ -501,12 +501,186 @@ pub fn ecmul(
     }
 }
 
-fn ecpairing(
-    _calldata: &Bytes,
-    _gas_for_call: u64,
-    _consumed_gas: &mut u64,
+pub fn ecpairing(
+    calldata: &Bytes,
+    gas_for_call: u64,
+    consumed_gas: &mut u64,
 ) -> Result<Bytes, VMError> {
-    Ok(Bytes::new())
+    // The input must always be a multiple of 192 (6 32-byte values)
+    if calldata.len() % 192 != 0 {
+        return Err(VMError::PrecompileError(PrecompileError::ParsingInputError));
+    }
+
+    let groups_number = calldata.len() / 192;
+
+    // Consume gas
+    let gas_cost = gas_cost::ecpairing(groups_number)?;
+    increase_precompile_consumed_gas(gas_for_call, gas_cost, consumed_gas)?;
+
+    let mut mul: FieldElement<Degree12ExtensionField> = QuadraticExtensionFieldElement::one();
+    for group_number in 0..groups_number {
+        // Define the group indexes and slice calldata to get the group data
+        let group_start = group_number
+            .checked_mul(192)
+            .ok_or(InternalError::ArithmeticOperationOverflow)?;
+        let next_group_start = (group_number
+            .checked_add(1)
+            .ok_or(InternalError::ArithmeticOperationOverflow)?)
+        .checked_mul(192)
+        .ok_or(InternalError::ArithmeticOperationOverflow)?;
+
+        let group_data = calldata
+            .get(group_start..next_group_start)
+            .ok_or(InternalError::SlicingError)?;
+
+        let first_point_x = group_data.get(..32).ok_or(InternalError::SlicingError)?;
+        let first_point_y = group_data.get(32..64).ok_or(InternalError::SlicingError)?;
+
+        // Infinite is defined by (0,0). Any other zero-combination is invalid
+        if (U256::from_big_endian(first_point_x) == U256::zero()
+            && U256::from_big_endian(first_point_y) != U256::zero())
+            || (U256::from_big_endian(first_point_x) != U256::zero()
+                && U256::from_big_endian(first_point_y) == U256::zero())
```

Comment:
Isn't this like an exclusive or?
I don't really know if this is equivalent but it might be:
```rust
if (U256::from_big_endian(first_point_x) == U256::zero())
    ^ (U256::from_big_endian(first_point_y) == U256::zero())
```

Reply:
Done in [Use xor in non valid infinity checking](https://github.com/lambdaclass/ethrex/pull/1560/commits/0f72780ba6d2203ae890f3956026011b7a8386e5)